keps/349-all-or-nothing/README.md

```diff
@@ -0,0 +1,482 @@
+
+My use case be supported by adding `kueue.x-k8s.io/startup-requirements: QuotaAndPodsReady`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources. However, I don't want to risk
+two Jobs deadlock in case of issues with provisioning nodes to satisfy the configured
+cluster queue quota. This could happen when the job corresponding to these Jobs don't
+specify priorities or specify the same priority.
+
+My use case be supported by adding `kueue.x-k8s.io/startup-requirements: QuotaAndPodsReady`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Choosing the API object to opt-in for the mechanism
+
+Let's review the options for the best place for the API and their pros&cons:
+- job (current proposal)
+- local queue
+- cluster queue
+- component
+
+**Workload/Job level pros**:
+- throughput - the same cluster queue may run workloads which don't need any guarantees and thus get admitted faster.
+- flexibility -  we can easily implement queue-level (or component-level) by translating the queue-level property into workload property during workload creation. For example, add the workload annotation to every workload created for a given local queue.
+- conceptually, requiring all pods to run at the same time is a property of a workload itself.
+
+**Workload/Job level cons**:
+- introducing a new annotation on a Job level will increase the complexity of the job reconciler which is a reference Job-based implementation of CRD <> Workload integration. Similarly, introducing
+  a timeout on reaching `PodsReady` condition could increase the complexity of the reference implementation.
+
+**Local Queue level pros**:
+- does not increase complexity of the Job reconciler which is the reference implementation for the CRD - Workload integration in case of Job,
+- allows for the same throughput as workload level if Jobs that don't require the `PodsReady` condition are sent to another LocalQueue
+- the workload-level field could be still populated based on the assigned local queue
+- could be a natual place to also define timeout for reaching `PodsReady`, which is not component-level
+
+**Local Queue level cons**:
+- similar complexity of implementation comparing to workload level - in order to ensure exclusive start of pods we still need to both block admission of new workloads until in `PodsReady` and await for all already workloads to be in the `PodsReady` condition, as some workloads might be executed on the cluster bypassing the queues.
+- throughput - it could be compensated by creating dedicated queues for workloads requiring all-or-nothing or not requiring it.
```

Comment:
this is not necessarily true, because once a workload that requires all-or-nothing is being admitted, we wait for all workloads in the cluster, don't we?

In any case, I think an API at LocalQueue level makes sense. But two things:
- You need to make sure to keep it up to date (a workload could be created before the queue, workloads can switch queues, etc).
- You need to make clear in the documentation that a non-empty LocalQueue with all-or-nothing will impact the entire cluster, not just the workloads in that queue.
-

Reply:
> this is not necessarily true, because once a workload that requires all-or-nothing is being admitted, we wait for all workloads in the cluster, don't we?

indeed, removed the misleading point.